File: ruby/pokemon/pokemon.rb
# Pokemonクラスの定義
class Pokemon
  attr_reader :name

  def initialize(name, type1, type2, hp)
    @name = name
    @type1 = type1
    @type2 = type2
    @hp = hp
  end

  def change_name(new_name)
    return "#{new_name} は名前に使用できません。" if new_name == '***'

    @name = new_name
  end

  def attack
    puts "#{@name} のこうげき！"
  end
end

# ピカチュウクラスの定義
class Pikachu < Pokemon
  def attack
    super
    puts "#{@name} の10万ボルト!"
  end
end

# ゼニガメクラスの定義
class Zenigame < Pokemon
  def attack
    super
    puts "#{@name} のみずてっぽう!"
  end
end

File: python/pokemon/pokemon.py
class Pokemon:
    def __init__(self, name, type1, type2, hp):
        self._name = name
        self._type1 = type1
        self._type2 = type2
        self._hp = hp

    def attack(self):
        print(f"{self._name}の攻撃!!!")

    def change_name(self, newName):
        if "F Word" in newName:
            print(f"{newName}は名前に使用できません。")
            return
        self._name = newName

    @property
    def name(self):
        return self._name


class Pikachu(Pokemon):
    def attack(self):
        super().attack()
        print(f"{self._name}の10万ボルト!!!")


class Zenigame(Pokemon):
    def attack(self):
        super().attack()
        print(f"{self._name}の水鉄砲!!!")


if __name__ == "__main__":
    pikachu = Pikachu("ピカチュウ", "electric", "", 10)
    pikachu.attack()
    pikachu.change_name("F Word ピカチュウ")
    print(pikachu.name)

    zenigame = Zenigame("ゼニガメ", "water", "", 30)
    zenigame.attack()
    zenigame.change_name("カメール")
    print(zenigame.name)
